### ISSUE TITLE
lxd high cpu (approx 200+ CPU% for extended time)

### ISSUE THREAD

**Issue by jameinel** [lxc/lxd#4025]
# Required information

 * Distribution: xenial
 * The output of "lxc info" or if that fails:
```
config:
  core.https_address: '[::]'
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses:
  - 192.168.0.106:8443
  - 10.67.99.1:8443
  - 192.168.122.1:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
  certificate_fingerprint: 7d043aebcbd0c5aa5d124fae3b38efd944ba53d41dc2da3342d4625a7ba120a5
  driver: lxc
  driver_version: 2.0.8
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-97-generic
  server: lxd
  server_pid: 27235
  server_version: "2.18"
  storage: zfs | btrfs
  storage_version: 0.6.5.6-0ubuntu16 | 4.4
```
# Issue description

LXD has gotten itself into some sort of loop, consuming lots of CPU. I've even tried restarting the daemon and it goes right back to 200% CPU.

# Steps to reproduce

I haven't been able to trigger this, I'm just in this state right now.

# Information to attach

I tried running: lxc monitor --type=logging output
```
metadata:
  context:
    ip: 10.67.99.222:34602
    method: GET
    url: /1.0/profiles?recursion=1
  level: dbug
  message: handling
timestamp: 2017-11-15T15:28:55.419137077+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.419085864+04:00
type: logging


metadata:
  context:
    ip: 10.67.99.222:34600
    method: GET
    url: /1.0/profiles?recursion=1
  level: dbug
  message: handling
timestamp: 2017-11-15T15:28:55.420008148+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.419968505+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.416812963+04:00
type: logging


metadata:
  context: {}
  level: dbug
  message: Initializing a ZFS driver.
timestamp: 2017-11-15T15:28:55.422764504+04:00
type: logging

metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.42099057+04:00
type: logging


metadata:
  context:
    ip: 10.67.99.222:34610
    method: GET
    url: /1.0
  level: dbug
  message: handling
timestamp: 2017-11-15T15:28:55.420954368+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.42091566+04:00
type: logging


metadata:
  context: {}
  level: dbug
  message: Initializing a BTRFS driver.
timestamp: 2017-11-15T15:28:55.453138406+04:00
type: logging


metadata:
  context:
    ip: 10.67.99.222:34616
    method: GET
    url: /1.0/containers?recursion=1
  level: dbug
  message: handling
timestamp: 2017-11-15T15:28:55.47418381+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.474140252+04:00
type: logging

metadata:
  context:
    ip: 10.67.99.222:34618
    method: GET
    url: /1.0/containers?recursion=1
  level: dbug
  message: handling
timestamp: 2017-11-15T15:28:55.473971429+04:00
type: logging

metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.473900273+04:00
type: logging


metadata:
  context: {}
  level: dbug
  message: Initializing a ZFS driver.
timestamp: 2017-11-15T15:28:55.515109833+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.513871322+04:00
type: logging


metadata:
  context:
    ip: 10.67.99.222:34598
    method: GET
    url: /1.0
  level: dbug
  message: handling
timestamp: 2017-11-15T15:28:55.513936972+04:00
type: logging


metadata:
  context:
    k: "4"
  level: dbug
  message: Found cert
timestamp: 2017-11-15T15:28:55.513982452+04:00
type: logging


metadata:
  context: {}
  level: dbug
  message: Initializing a BTRFS driver.
timestamp: 2017-11-15T15:28:55.530193391+04:00
type: logging
```

It looks like its continually resetting the cert/ZFS/BTRFS drivers?

It did eventually settle down, but it lasted 5-10 minutes (long enough for me to try restarting, and filing this bug.)

**Comment by jameinel** [comment]
This might be something Juju is doing.
I was playing around with a Juju controller that was using LXD, and had 200 models (but no actual instances running).
Is it possible that each model ends up connecting to the LXD agent independently, and all of those connections end up causing problems?

I see what appears to be the same issue as soon as I restart the Juju controller.

[lxd-monitor.txt.gz](https://github.com/lxc/lxd/files/1474529/lxd-monitor.txt.gz)

**Comment by stgraber** [comment]
How many containers do you have?

**Comment by jameinel** [comment]
I only have 1 container. but I have 200 models in Juju that haven't
deployed anything yet.

**Comment by stgraber** [comment]
You've done 158 API queries over a period of 17s, some of those calls, especially listing containers with all properties (/1.0/containers?recursion=1" can be pretty expensive as that needs to do quite a lot of DB and disk accesses.

**Comment by stgraber** [comment]
Nowhere near as bad as accessing /state which you don't seem to be doing particularly often, so I'm still surprised that it'd cause that much load...

**Comment by jameinel** [comment]
So it isn't anything like a common use case. I'm guessing that we run 1
Provisioner per model, which wants to track what containers are running
etc. So it is taking inventory, etc. And if you end up with 200 models,
they all do inventory on your local LXD agent. Though it still seems like
if I'm seeing Minutes at 200, we should be doing something so that it is
<second with a handful of models.

If there are any queries that we should be avoiding, please let us know. I
have a strong feeling we did this because we got it to work, and not so
much that we are positive we're not doing more than we should.

**Comment by stgraber** [comment]
Querying /1.0/containers?recursion=1 (GetContainers() in the current Go API) can get pretty expensive on some systems. I'm working on a branch that should significantly reduce that overhead now but it still doesn't feel like something that juju has a good reason to keep doing in the background unless it's actually doing something.

**Comment by jameinel** [comment]
This seems to happen for a couple minutes on a server restart and then we
go quiet again. And likely I only made it this bad by having a 200x
multiplier. but I'll try to have a look for what we're doing with that
information
would you expect it to be that expensive with only a single container?

John
=:->

**Comment by jameinel** [comment]
Just to state, the function is actually:
func (c *Client) ListContainers() ([]api.Container, error) {
    if c.Remote.Public {
        return nil, fmt.Errorf("This function isn't supported by public remotes.")
    }

    resp, err := c.get("containers?recursion=1")
...

Note that Juju's call actually wants a filtered list of instances. Namely:
func (client *instanceClient) Instances(prefix string, statuses ...string) ([]Instance, error) {
    infos, err := client.raw.ListContainers()
    if err != nil {
        return nil, errors.Trace(err)
    }

    var insts []Instance
    for _, info := range infos {
        name := info.Name
        if prefix != "" && !strings.HasPrefix(name, prefix) {
            continue
        }
...

So ideally we would only be calling ListContainers on the specific containers that are in a given model by passing in some sort of filtering.
In my test case, I actually only really had a very small number of containers (I believe it was under 10, in my initial test I had only a very single container, but I'm not sure what the load was there.)

Again, it isn't a standard use case for a Juju user to be creating lots of models on the local provider, but we do generally need to be able to find the instances that we have running for a particular model during startup.
10min for 200 calls is still 3s per call, which seems rather high. I'll play around a bit and see if that is the specific amount, or if there is some other bug inside our code base that is actually causing us to call this more than once per model.
Still it would be ideal if we could pass in some sort of filtering into the List request, and have them filtered out server-side, instead of getting all of the container information and then filtering it out client side.

**Comment by jameinel** [comment]
Note that it doesn't seem to *just* be this call that is a problem.
On that same machine with just a single container running, I can see that making requests over the HTTPS port is about 10x faster than doing it over the unix socket, but either way, I can still do:
$ time (for j in `seq 10`; do for i in `seq 10`; do curl --insecure --cert /home/jameinel/dev/tmp/client.cert --key ~/dev/tmp/client.key https://localhost:8443/1.0/containers?recursive=1 & done; time wait; done)
...
real    0m6.464s
user    0m14.432s
sys     0m1.652s

so that is 100 requests done in parallel, and it still only takes 6s to complete. So an average of 600ms and it supports doing them significantly in parallel.
And even with all of those queries, I don't really see the LXD cpu load go high. While when I'm doing:
`juju add-model` 10 at a time, I see the CPU going up to 190-250% and staying there for an extended period of time.

Once I've done that 'add-model', if I go back to curl ... containers?recursive=1 I only see the load go to 110% for 10 parallel requests. Even if I go up to 50 parallel requests, I only see load topping out at 100-140%.

I also tried checking if the problem was "profiles?recursive=1" but that takes essentially exactly the same amount of time. (approx  6.8s to issue 10x10 requests.)

Is there something weird about "initializing ZFS driver" and "initializing BTRFS driver" that we're doing wrong causing LXD to re-init all the time?

Something specifically is fishy, because with just curl to /containers and /profiles I'm never able to go above 100% cpu on lxd, but I am able to go higher w/ juju.

**Comment by stgraber** [comment]
Can you reasonably easily test this against the LXD snap from the edge channel?

After you first reported this issue I did the work to update a bunch of internal calls within our container management code to only initialize the disk driver and other disk-related bits if and only if we absolutely need to.

This should very significantly cut down on those "initializing ZFS driver" and "initializing BTRFS driver"  messages, though we've not seen any significant performance improvements as a result of this in our daily performance tests.

**Comment by jameinel** [comment]
As far as I can tell, it didn't actually change the *number* of initializing messages. I attached a log of restarting the controller with ~200 models.

What was *noticeably* better was how long it took before LXD quieted down. I think it used to be at least 10 minutes, and this time was probably under 2minutes. (It often was under 200%, but did peak around 240% so the load was still reasonably high.)

[logging.txt.gz](https://github.com/lxc/lxd/files/1565546/logging.txt.gz)

**Comment by stgraber** [comment]
@jameinel so looks like it did actually help for the /1.0/containers calls, those don't appear to show those messages anymore. But it shows that "GET /1.0" still causes all the storage backends to be loaded.

I'll take a look at the /1.0 code to see if we can make it so we don't load every single one of the storage pools there. For well behaved clients that's not a huge deal as GET /1.0 is a one-time action when loading the client, but in your case it looks like you're creating a LOT of clients causing it to get hammered.

**Comment by stgraber** [comment]
```
stgraber@castiana:~/Downloads$ zgrep '/1.0$' logging.txt.gz | wc -l
1024
stgraber@castiana:~/Downloads$ zgrep message logging.txt.gz  | sort | uniq -c 
   3243   message: Found cert
   2219   message: handling
   1024   message: Initializing a BTRFS driver.
   1025   message: Initializing a DIR driver.
      1   message: 'New events listener: 9a5c6806-5b2d-4eba-9387-d0f520a1e0d5'
```

**Comment by stgraber** [comment]
So if we can get rid of those calls, that should get the number of storage calls down to 0 in your case.

**Comment by stgraber** [comment]
@jameinel can you refresh the edge snap and try again? I wonder how much of a difference this change made for you.

**Comment by jameinel** [comment]
snap list now says:
lxd            git-3effa25          5328  canonical  -

Time for Juju to be restarted with ~180 models is right about 2 minutes.
Attached is the logging.txt.gz

Restart speed seems approximately the same as my last test. I do think I
see far fewer storage initialization calls. But maybe the "Found cert" or
"/1.0/containers?recursion=1" is being slow.

BTW, there doesn't seem to be a way to list a subset of the containers, is
there?

If you had more than 1 model on LXD, we tag the model-uuid on each
container. Which means that each model worker could search for only its
containers, rather than getting all containers and then filtering it out on
the tags of the returned list.

As I mentioned, this isn't a standard use case for Juju and LXD, this is
more scale testing/stress testing. (If the average useful model has about 5
machines, you aren't really going to have 200 models == 1000 containers on
a single machine.)
However, as LXD grows clustering, maybe small hundreds starts to make more
sense, so knowing where the pain points are is probably useful.

John
=:->

**Comment by jameinel** [comment]
This is with the edge snap (seems it didn't get attached properly)
[logging.txt.gz](https://github.com/lxc/lxd/files/1570441/logging.txt.gz)

**Comment by stgraber** [comment]
```
stgraber@castiana:~/Downloads$ zcat logging\ \(1\).txt.gz | grep "message" | grep -v "handling" | sort | uniq -c
   3966   message: Found cert
      1   message: Initializing a DIR driver.
      1   message: 'New events listener: 970bcecc-7b15-43d5-9af8-a4e6718c4305'
stgraber@castiana:~/Downloads$ zcat logging\ \(1\).txt.gz | grep "url:" | sort | uniq -c
   1262     url: /1.0
      1     url: /1.0/containers/juju-80f953-0/state
    179     url: /1.0/containers?recursion=1
    631     url: /1.0/profiles/default
    631     url: /1.0/profiles?recursion=1
```

So yeah, the storage init messages are all gone now.

**Comment by stgraber** [comment]
The "found cert" messages are very cheap, we have an in-memory cache of all the trusted certificates so don't even hit the database for those. There's no subprocesses being spawned for those either.

**Comment by jameinel** [comment]
Doing another restart with logging. The first message triggers at:
  message: Found cert
timestamp: 2017-12-19T05:03:57.090009814Z
  message: Found cert
timestamp: 2017-12-19T05:05:54.03801231Z

So that gives approx 1m57s to restart and reach quiescence.

I then did a refresh back to stable, and got:
lxd            2.20                 5235  canonical  -

With stable the first and last messages were:
  message: Initializing a DIR driver.
timestamp: 2017-12-19T05:09:50.041492741Z
...
  message: Found cert
timestamp: 2017-12-19T05:12:10.875215005Z

Which is roughly 2min20s.

**Comment by stgraber** [comment]
Ok, so better and less noisy but clearly the system is still quite busy processing all the requests.

For your other question, there isn't a way to get just a subset of the containers. You could in theory call GetContainerNames() which would get you just the names and then call GetContainer() for each of the ones you care about but that'd require you to keep track of the names. I also suspect this would actually lead to more load and worse performance in general.

The very expensive call is GetContainerState() as that involves a lot of kernel calls to the container.
GetContainers() is 99% database data and 1% unix socket communication with the container's monitor process, given that the database is likely to be fully cached in memory, I'd expect most of the time to actually be spent in TLS session handling.

**Comment by jameinel** [comment]
As a test I used edge + a quick hacked juju that lists the names first, and then iterates those names for containers that are relevant and then calls GetContainerInfo on them.

The new messages are at:
  message: Found cert
timestamp: 2017-12-19T06:59:07.121641077Z
  message: Found cert
timestamp: 2017-12-19T07:01:07.563187779Z

[logging-names.txt.gz](https://github.com/lxc/lxd/files/1570666/logging-names.txt.gz)

That seems to say the time is exactly 2min. I'm guessing the issue is very much as you say with things like TLS, etc.
